src/IB/BrinkmanPenalizationRigidBodyDynamics.cpp

```diff
@@ -588,15 +524,500 @@ BrinkmanPenalizationRigidBodyDynamics::getFromRestart()
+                        const double mu = 0.5 * (mu_lower + mu_upper);
+                        penalty_mu_scale = mu / (h_min * h_min);
+                    }
+                    const double penalty = d_penalty_factor * (penalty_rho_scale + penalty_mu_scale);
+
+                    (*u_in_data)(s_i) = d_trans_vel_new(axis) + Wxdr(axis);
+                    (*u_in_data)(s_i) *= (1.0 - Hphi) * penalty;
+                }
+
+                // Near the solid interface impose different penalty factors to normal
+                // and tangential velocity components
+                if (phi <= alpha && phi >= -alpha)
+                {
+                    const Eigen::Vector3d r = IBTK::IndexUtilities::getSideCenter<Eigen::Vector3d>(*patch, s_i);
+                    const Eigen::Vector3d dr = r - d_center_of_mass_new;
+                    const Eigen::Vector3d Wxdr = d_rot_vel_new.cross(dr);
+                    const Eigen::Vector3d Ub = d_trans_vel_new + Wxdr;
+
+                    // Compute normal at the side center from the solid level set. Interpolate
+                    // tangential components of velocity at the side center. We treat tangential
+                    // components of the velocity explicit in time
+                    Eigen::Vector3d n = Eigen::Vector3d::Zero();
+                    Eigen::Vector3d Ut = Eigen::Vector3d::Zero();
+                    for (unsigned int d = 0; d < NDIM; ++d)
+                    {
+                        if (d == axis)
+                        {
+                            n(axis) = (phi_upper - phi_lower) / patch_dx[axis];
+                            Ut(axis) = 0.0;
+                        }
+                        else
+                        {
+                            CellIndex<NDIM> offset(0);
+                            offset(d) = 1;
+
+                            const double h = patch_dx[d];
+                            n(d) = ((*ls_solid_data)(c_l + offset) - (*ls_solid_data)(c_l)) +
+                                   ((*ls_solid_data)(c_u + offset) - (*ls_solid_data)(c_u)) +
+                                   ((*ls_solid_data)(c_l) - (*ls_solid_data)(c_l - offset)) +
+                                   ((*ls_solid_data)(c_u) - (*ls_solid_data)(c_u - offset));
+                            n(d) *= 1.0 / (4.0 * h);
+
+                            Ut(d) = (*u_new_data)(SideIndex<NDIM>(c_u, d, SideIndex<NDIM>::Lower)) +
+                                    (*u_new_data)(SideIndex<NDIM>(c_u, d, SideIndex<NDIM>::Upper)) +
+                                    (*u_new_data)(SideIndex<NDIM>(c_l, d, SideIndex<NDIM>::Lower)) +
+                                    (*u_new_data)(SideIndex<NDIM>(c_l, d, SideIndex<NDIM>::Upper));
+                            Ut(d) *= 0.25;
+                        }
+                    }
+                    const double n_norm = n.norm();
+                    if (!IBTK::rel_equal_eps(n_norm, 0.0))
```

Comment:
Consider using `abs_equal_eps` to compare numbers with zero. See
https://github.com/IBAMR/IBAMR/blob/86c3a9f360cebcb372ed77046dae5fe75b6187bf/tests/IBTK/equal_eps.output